Define chain rules for complex functions (#167)
JuliaDiff/ChainRulesCore.jl: docs/src/FAQ.md

This means your package can depend on the light-weight ChainRulesCore.jl, and make ChainRulesTestUtils.jl a test-only dependency.

Remember to read the section on [On writing good `rrule` / `frule` methods](@ref).

## How do chain rules work for complex functions?

> `ChainRules.jl` follows the convention that `frule` applied to a function ``f(x + i y) = u(x,y) + i v(x,y)`` with perturbation ``\Delta x + i \Delta y`` returns the value and

Reviewer: Super pedantic, but elsewhere, "ChainRules.jl" is not in fixed-width font.
```suggestion
ChainRules.jl follows the convention that `frule` applied to a function ``f(x + i y) = u(x,y) + i v(x,y)`` with perturbation ``\Delta x + i \Delta y`` returns the value and
```

Author: :+1: 
Also removed the .jl.

\tfrac{\partial v}{\partial x} & \tfrac{\partial v}{\partial y} \\
\end{pmatrix}
\begin{pmatrix}
\Delta x \\ \Delta y
\end{pmatrix}

```
and `rrule` corresponds to
```math
\begin{pmatrix}
\Delta u \, \tfrac{\partial u}{\partial x} + \Delta v \, \tfrac{\partial v}{\partial x}
&
\Delta u \, \tfrac{\partial u}{\partial y} + \Delta v \, \tfrac{\partial v}{\partial y}
\end{pmatrix}
=
\begin{pmatrix}
\Delta u & \Delta v
\end{pmatrix}
\begin{pmatrix}
\tfrac{\partial u}{\partial x} & \tfrac{\partial u}{\partial y} \\
\tfrac{\partial v}{\partial x} & \tfrac{\partial v}{\partial y} \\
\end{pmatrix}
.
```
The Jacobian of ``f:\mathbb{C} \to \mathbb{C}`` interpreted as a function ``\mathbb{R}^2 \to \mathbb{R}^2`` can hence be evaluated using either of the following functions.
```
function jacobian_via_frule(f,z)
    fz,df_dx = frule((Zero(), 1),f,z)
    fz,df_dy = frule((Zero(),im),f,z)
    return [
        real(df_dx)  real(df_dy)
        imag(df_dx)  imag(df_dy)
    ]
end
```
```
function jacobian_via_rrule(f,z)
    fz, pullback = rrule(f,z)
    _,du_dz = pullback( 1)
    _,dv_dz = pullback(im)
    return [
        real(du_dz)  imag(du_dz)
        real(dv_dz)  imag(dv_dz)
    ]
end
```

> If ``f(z)`` is holomorphic, then the derivative part of `frule` can be implemented as ``f'(z) \, \Delta z`` and the derivative part of `rrule` can be implemented as ``\Delta f \,  \overline{f'(z)}``.

Reviewer 2: This is incredibly pedantic, but the following would make this line up slightly better with the multidimensional case below:
```suggestion
If ``f(z)`` is holomorphic, then the derivative part of `frule` can be implemented as ``f'(z) \, \Delta z`` and the derivative part of `rrule` can be implemented as ``\overline{f'(z)} \, \Delta f``.
```

Reviewer: Also, can you indicate here that `\overline{.}` means complex conjugate? Elsewhere we use `'` to mean conjugate and `\overline` to mean "adjoint".

Alternatively, since `\overline` is only used here several times for conjugation, perhaps it would be better to use `\operatorname{conj}`?

Author: Changed to `\operatorname{conj}`. This is probably the least confusing regardless of your background.

Consequently, holomorphic derivatives can be evaluated using either of the following functions.
```
function holomorphic_derivative_via_frule(f,z)
    fz,df_dz = frule((Zero(),1),f,z)
    return df_dz
end
```
```
function holomorphic_derivative_via_rrule(f,z)
    fz, pullback = rrule(f,z)
    dself, conj_df_dz = pullback(1)
    return conj(conj_df_dz)
end
```

In the multi-dimensional case ``f : \mathbb{C}^m \to \mathbb{C}^n``, the derivative part of `frule` is given by
```math
\sum_{\ell = 1}^m \tfrac{\partial u_k}{\partial x_\ell} \, \Delta x_\ell + \tfrac{\partial u_k}{\partial y_\ell} \, \Delta y_\ell + i \, \Bigl( \tfrac{\partial v_k}{\partial x_\ell} \, \Delta x_\ell + \tfrac{\partial v_k}{\partial y_\ell} \, \Delta y_\ell \Bigr)
\qquad\text{for } k = 1, \ldots, n
```
and the derivative part of `rrule` is given by
```math
\sum_{k = 1}^n \Delta u_k \, \tfrac{\partial u_k}{\partial x_\ell} + \Delta v_k \, \tfrac{\partial v_k}{\partial x_\ell} + i \, \Bigl(\Delta u_k \, \tfrac{\partial u_k }{\partial y_\ell} + \Delta v_k \, \tfrac{\partial v_k}{\partial y_\ell} \Bigr)
\qquad\text{for } \ell = 1, \ldots, m
```

> In particular, if ``f(z)`` is holomorphic (i.e. every component function ``f_k(z)`` is holomorphic in every component ``z_\ell`` of the argument), then the derivative part of `frule` can be implemented as ``\nabla f(z) \, \Delta z`` and the derivative part of `rrule` can be implemented as ``\overline{\nabla f(z)}^H \, \Delta f``, where ``\nabla f(z)`` denotes the matrix of holomorphic derivatives ``(\nabla f)_{k\ell} = \frac{\partial f_k}{\partial z_\ell}`` and ``A^H = \overline{A}^T`` denotes the Hermitian transpose.

Reviewer: In `\overline{\nabla f(z)}^H`, does the overline indicate conjugation, and if so, isn't it then redundant?

Author: Indeed, the extra `\overline` was wrong. Fixed now.

This means your package can depend on the light-weight ChainRulesCore.jl, and make ChainRulesTestUtils.jl a test-only dependency.

Remember to read the section on [On writing good `rrule` / `frule` methods](@ref).

> ## Chain rules for complex functions

Reviewer 3: ```suggestion
## How do chain rules work for complex functions?
```

`ChainRules.jl` follows the convention that `frule` applied to a function `f(x + im*y) = u(x,y) + im*v(x,y)` with perturbation `ẋ + im*ẏ` returns
```
∂u/∂x * ẋ + ∂u/∂y * ẏ + im * ( ∂v/∂x * ẋ    ∂v/∂y * ẏ )
```

> and similarly `rrule` applied to the same function with adjoint `ū + im*v̄` returns

Reviewer 2: ```suggestion
and similarly `rrule` applied to the same function evaluates the function and return a pullback which, when applied to adjoint `ū + im*v̄`, returns
```

Author: I get that my version wasn't 100% accurate, but I think it is justified to be slightly imprecise here for the sake of getting straight to the point. 

If we do make this change than we should also say that `frule` returns the function value and the derivative (cf. first half of sentence).

Reviewer 4: FWIW, i prefer accuracy over brevity. I agree it'd be good to be consistent in the level of detail for both `frule` and `rrule`

Author: Most people seem to agree with you, so I guess that's what we're sticking with.

```
ū * ∂u/∂x + v̄ * ∂v/∂x + im * ( ū * ∂u/∂y + v̄ * ∂v/∂y )
```

> Note that these rules can be interpreted as multiplication with the Jacobian and transposed Jacobian, respectively of `(x,y)->[u(x,y), v(x,y])`. 

Reviewer 2: @ettersi could you provide an expression for this interpretation directly please? i.e. write out the jvp and vjp in matrix-vector form?

Author: Sure. How do I insert a display math equation?

Author: Done

> `ChainRules.jl` follows the convention that `frule` applied to a function `f(x + im*y) = u(x,y) + im*v(x,y)` with perturbation `Δx + im*Δy` returns
> ```
> ∂u/∂x * Δx + ∂u/∂y * Δy + im * ( ∂v/∂x * Δx    ∂v/∂y * Δy )
> ```
> and similarly `rrule` applied to the same function with adjoint `ū + im*v̄` returns
> ```
> ū * ∂u/∂x + v̄ * ∂v/∂x + im * ( ū * ∂u/∂y + v̄ * ∂v/∂y )
> ```
> Note that these rules can be interpreted as multiplication with the Jacobian and transposed Jacobian, respectively of `(x,y)->[u(x,y), v(x,y])`. 
> 
> If `f(z)` is holomorphic, then `frule` can be implemented as `f'(z) * Δz` and `rrule` can be implemented as `f̄ * conj(f'(z))`. 

Reviewer 2: ```suggestion
`ChainRules.jl` follows the convention that `frule` applied to a function `f(x + im*y) = u(x,y) + im*v(x,y)` with perturbation `ẋ + im*ẏ` returns
```
∂u/∂x * ẋ + ∂u/∂y * ẏ + im * ( ∂v/∂x * ẋ    ∂v/∂y * ẏ )
```
and similarly `rrule` applied to the same function with adjoint `ū + im*v̄` returns
```
ū * ∂u/∂x + v̄ * ∂v/∂x + im * ( ū * ∂u/∂y + v̄ * ∂v/∂y )
```
Note that these rules can be interpreted as multiplication with the Jacobian and transposed Jacobian, respectively of `(x,y)->[u(x,y), v(x,y])`. 

If `f(z)` is holomorphic, then `frule` can be implemented as `f'(z) * ż` and `rrule` can be implemented as `f̄ * conj(f'(z))`. 
```

Reviewer 2: Didn't render properly, but just changes `Δx` to `ẋ` etc.

Author: On my system (Chrome on Ubuntu), the dot is quite hard to spot. Also, `v̄` doesn't render properly for me. Maybe it's better to avoid these diacritics altogether, but of course that's your call.

This means your package can depend on the light-weight ChainRulesCore.jl, and make ChainRulesTestUtils.jl a test-only dependency.

Remember to read the section on [On writing good `rrule` / `frule` methods](@ref).

## How do chain rules work for complex functions?

> `ChainRules.jl` follows the convention that `frule` applied to a function ``f(x + i y) = u(x,y) + i \, v(x,y)`` with perturbation ``\dot x + i \dot y`` returns

Reviewer 3: ```suggestion
`ChainRules.jl` follows the convention that `frule` applied to a function ``f(x + i y) = u(x,y) + i \, v(x,y)`` with perturbation ``\dot x + i \dot y`` returns the value and 
```

> `ChainRules.jl` follows the convention that `frule` applied to a function ``f(x + i y) = u(x,y) + i \, v(x,y)`` with perturbation ``\dot x + i \dot y`` returns the value and 

Reviewer 5: I think that is too general of a statement. The way you describe it only really applies for `frule` if your input is real or `rrule` if your eventual output is real. I find it quite useful to think of forward-mode AD as propagating a perturbation in the input and getting out a derivative wrt to that input and reverse-mode as propagating a sensitivity in the output and getting out a vector of directional derivatives. Those perturbations/sensitivities can be complex, but not the input/output itself.

Author: The point of this FAQ is to point out two things:

1) `frule` and `rrule` are always defined (e.g. even for non-holomorphic `Complex -> Complex` functions). 
2) These definitions are sufficient to compute all quantities of interest. 

I've added a bit more details to the documentation, I hope this helps to clarify.

Author: I guess your comment arose because this convention for complex `frule` and `rrule` leads to a really awkward API for e.g. extracting Jacobians. This is definitely an issue, but the right place to fix this is probably in another package built on top of `ChainRules`, not in `ChainRules` itself.

Reviewer 5: > These definitions are sufficient to compite all quantities of interest.

Depends on what you are interested in. Like I said above, only if the input/output is real will you get a meaningful result. For `C -> C` you would either want the Wirtinger primal and conjugate or a 2x2 Jacobian, which we don't currently support. We should clarify that, so that people don't get any wrong impressions.

Author: I have added a note at the very end which I hope addresses your concern. Let me know what you think.

```math
\tfrac{\partial u}{\partial x} \, \dot x + \tfrac{\partial u}{\partial y} \, \dot y + i \, \Bigl( \tfrac{\partial v}{\partial x} \, \dot x + \tfrac{\partial v}{\partial y} \, \dot y \Bigr)
,
```

> and similarly `rrule` applied to the same function evaluates the function and returns a pullback which, when applied to adjoint ``\bar u + i \bar v``, returns

Reviewer 5: I think `\bar u` is a bit confusing notation here, because it's usually used for conjugation. What do you think of `\mathrm{d}u` instead, since you are basically pulling back differential one-forms? The notation should probably also be explained briefly.

Author: Good point! I replaced both `\bar` and `\dot` with `\Delta`. This notation (and the earlier `dot` and  `bar` notation) are documented further up on this FAQ page.

Reviewer 5: I don't think it makes sense to use `\Delta` for both, `frule` and `rrule`, because one lives in the tangent plane, whereas the other one lives in the cotangent plane.

Author: It's the notation proposed by the `ChainRules` documentation. Waiting for input from others.

.
```
Note that if we interpret complex numbers as vectors in ``\mathbb{R}^2``, then these rules correspond to left- and right-multiplication with the Jacobian of ``f(z)``, i.e. `frule` corresponds to 
```math
\begin{pmatrix}
\tfrac{\partial u}{\partial x} \, \dot x + \tfrac{\partial u}{\partial y} \, \dot y 
\\ 
\tfrac{\partial v}{\partial x} \, \dot x + \tfrac{\partial v}{\partial y} \, \dot y 
\end{pmatrix}
=
\begin{pmatrix}
\tfrac{\partial u}{\partial x} & \tfrac{\partial u}{\partial y} \\
\tfrac{\partial v}{\partial x} & \tfrac{\partial v}{\partial y} \\
\end{pmatrix}
\begin{pmatrix}
\dot x \\ \dot y
\end{pmatrix}

```
and `rrule` corresponds to 
```math
\begin{pmatrix}
\bar u \, \tfrac{\partial u}{\partial x} + \bar v \, \tfrac{\partial v}{\partial x} 
& 
\bar u \, \tfrac{\partial u}{\partial y} + \bar v \, \tfrac{\partial v}{\partial y} 
\end{pmatrix}
=
\begin{pmatrix}
\bar u & \bar v
\end{pmatrix}
\begin{pmatrix}
\tfrac{\partial u}{\partial x} & \tfrac{\partial u}{\partial y} \\
\tfrac{\partial v}{\partial x} & \tfrac{\partial v}{\partial y} \\
\end{pmatrix}
.
```

> If ``f(z)`` is holomorphic, then `frule` can be implemented as ``f'(z) \, \dot z`` and `rrule` can be implemented as ``\bar f \,  \overline{f'(z))}``. 

Reviewer 5: Since every non-trivial holomorphic function is `C -> C`, I think it's important to emphasize here that this is valid for intermediate steps, but if your input/output (depending on whether it's f-/rrule) is not real, this doesn't really give you a meaningful answer on its own.

Author: Correct. As pointed out above, we should provide another layer on top of `ChainRules` to provide a more reasonable API for evaluating Wirtinger derivatives, Jacobians, gradients of functions `Complex -> Real`, etc.

.
```
Similarly, `rrule` applied to the same function evaluates the function and returns a pullback which, when applied to the adjoint ``\Delta u + i \Delta v``, returns
```math
\Delta u \, \tfrac{\partial u}{\partial x} + \Delta v \, \tfrac{\partial v}{\partial x} + i \, \Bigl(\Delta u \, \tfrac{\partial u }{\partial y} + \Delta v \, \tfrac{\partial v}{\partial y} \Bigr)
.
```
If we interpret complex numbers as vectors in ``\mathbb{R}^2``, then these rules correspond to left- and right-multiplication with the Jacobian of ``f(z)``, i.e. `frule` corresponds to
```math
\begin{pmatrix}
\tfrac{\partial u}{\partial x} \, \Delta x + \tfrac{\partial u}{\partial y} \, \Delta y
\\
\tfrac{\partial v}{\partial x} \, \Delta x + \tfrac{\partial v}{\partial y} \, \Delta y
\end{pmatrix}
=
\begin{pmatrix}
\tfrac{\partial u}{\partial x} & \tfrac{\partial u}{\partial y} \\
\tfrac{\partial v}{\partial x} & \tfrac{\partial v}{\partial y} \\
\end{pmatrix}
\begin{pmatrix}
\Delta x \\ \Delta y
\end{pmatrix}

```

> and `rrule` corresponds to

Author: I assume you mean the (LaTeX) matvec formula for `rrule` here. I like the current right-multiplication interpretation of `rrule` since that's how I think about reverse differentiation, but you have a point as well. Someone just needs to make a call on this.

> and `rrule` corresponds to

Reviewer 2: Hmm good question. I think probably the right thing to do is to look at what we compute in practice. If I'm not mistaken, we compute `J^T v` so vote that. That way around would also be more consistent with the multi-dimensional case below.

> and `rrule` corresponds to

Reviewer: I agree. Though the vjp notation is more consistent with how we usually write the math, it's transposed from the code and risks introducing bugs.

> and `rrule` corresponds to

Author: "Follow the code structure" (rather than "follow the maths structure") makes sense to me. Updated this.

```math
\begin{pmatrix}
\Delta u \, \tfrac{\partial u}{\partial x} + \Delta v \, \tfrac{\partial v}{\partial x}
&
\Delta u \, \tfrac{\partial u}{\partial y} + \Delta v \, \tfrac{\partial v}{\partial y}
\end{pmatrix}
=
\begin{pmatrix}
\Delta u & \Delta v
\end{pmatrix}
\begin{pmatrix}
\tfrac{\partial u}{\partial x} & \tfrac{\partial u}{\partial y} \\
\tfrac{\partial v}{\partial x} & \tfrac{\partial v}{\partial y} \\
\end{pmatrix}
.
```
The Jacobian of ``f:\mathbb{C} \to \mathbb{C}`` interpreted as a function ``\mathbb{R}^2 \to \mathbb{R}^2`` can hence be evaluated using
```
function jacobian_via_frule(f,x)

>     fx,df_dx = frule((nothing, 1),f,x)

Author: Absolutely, that was confusing.

>     fx,df_dx = frule((nothing, 1),f,x)

Author: Changed as `Zero()` seems to be more standard in `ChainRules`. I kind of like `nothing` though as I feel it more clearly expresses that this argument shouldn't be there in the first place.

    fx,df_dy = frule((nothing,im),f,x)
    return [
        real(df_dx)  real(df_dy)
        imag(df_dx)  imag(df_dy)
    ]
end
```
```
function jacobian_via_rrule(f,x)
    fx, pullback = rrule(f,x)[2]
    _,du_dz = pullback( 1)
    _,dv_dz = pullback(im)
    return [
        real(du_dz)  imag(du_dz)
        real(dv_dz)  imag(dv_dz)
    ]
end
```

If ``f(z)`` is holomorphic, then the derivative part of `frule` can be implemented as ``f'(z) \, \Delta z`` and the derivative part of `rrule` can be implemented as ``\Delta f \,  \overline{f'(z)}``.
Consequently, holomorphic derivatives can be evaluated using
```
function holomorphic_derivative_via_frule(f,x)
    fx,df_dz = frule((nothing,1),f,x)
    return df_dz
end
```
```
function holomorphic_derivative_via_rrule(f,x)

>     fx, pullback = rrule(f,x)[2]

Author: Correct, these were typos.

This means your package can depend on the light-weight ChainRulesCore.jl, and make ChainRulesTestUtils.jl a test-only dependency.

Remember to read the section on [On writing good `rrule` / `frule` methods](@ref).

## How do chain rules work for complex functions?

ChainRules follows the convention that `frule` applied to a function ``f(x + i y) = u(x,y) + i v(x,y)`` with perturbation ``\Delta x + i \Delta y`` returns the value and
```math
\tfrac{\partial u}{\partial x} \, \Delta x + \tfrac{\partial u}{\partial y} \, \Delta y + i \, \Bigl( \tfrac{\partial v}{\partial x} \, \Delta x + \tfrac{\partial v}{\partial y} \, \Delta y \Bigr)
.
```

> Similarly, `rrule` applied to the same function returns the value and a pullback which, when applied to the adjoint ``\Delta u + i \Delta v``, returns

Reviewer 6: ```suggestion
Similarly, `rrule` applied to the same function returns the value and a pullback function which, when applied to the adjoint ``\Delta u + i \Delta v``, returns
```
I think it's good to emphasize explicitly that the `pullback` is a function.

```math
\Delta u \, \tfrac{\partial u}{\partial x} + \Delta v \, \tfrac{\partial v}{\partial x} + i \, \Bigl(\Delta u \, \tfrac{\partial u }{\partial y} + \Delta v \, \tfrac{\partial v}{\partial y} \Bigr)
.
```

> If we interpret complex numbers as vectors in ``\mathbb{R}^2``, then these rules correspond to multiplication with the (transposed) Jacobian of ``f(z)``, i.e. `frule` corresponds to

Reviewer 6: ```suggestion
If we interpret complex numbers as vectors in ``\mathbb{R}^2``, then the `frule` (`rrule`) corresponds to multiplication with the (transposed) Jacobian of ``f(z)``, i.e. `frule` corresponds to
```

Author: :+1:, but "the `frule`" sounds weird to my (nonnative) ear. Removed the "the".

```math
\begin{pmatrix}
\tfrac{\partial u}{\partial x} \, \Delta x + \tfrac{\partial u}{\partial y} \, \Delta y
\\
\tfrac{\partial v}{\partial x} \, \Delta x + \tfrac{\partial v}{\partial y} \, \Delta y
\end{pmatrix}
=
\begin{pmatrix}
\tfrac{\partial u}{\partial x} & \tfrac{\partial u}{\partial y} \\
\tfrac{\partial v}{\partial x} & \tfrac{\partial v}{\partial y} \\
\end{pmatrix}
\begin{pmatrix}
\Delta x \\ \Delta y
\end{pmatrix}

```
and `rrule` corresponds to
```math

> \begin{pmatrix}
> \tfrac{\partial u}{\partial x} \, \Delta u + \tfrac{\partial v}{\partial x} \, \Delta v
> \\
> \tfrac{\partial u}{\partial y} \, \Delta u + \tfrac{\partial v}{\partial y} \, \Delta v
> \end{pmatrix}
> =
> \begin{pmatrix}
> \tfrac{\partial u}{\partial x} & \tfrac{\partial v}{\partial x} \\
> \tfrac{\partial u}{\partial y} & \tfrac{\partial v}{\partial y} \\
> \end{pmatrix}
> \begin{pmatrix}
> \Delta u \\ \Delta v
> \end{pmatrix}

Reviewer 6: > Doesn't this address the question about the conjugate? It's only the scalar case, but I guess it is clear that if we conjugate scalars than we also conjugate matrices.

Sorry, I had looked, but I missed that line. I think in that case, the Jacobians shown aren't right.

I think the above lines should actually be
```suggestion
\begin{pmatrix}
conj\left(\tfrac{\partial u}{\partial x}\right) \, \Delta u + conj\left(\tfrac{\partial v}{\partial x}\right)  \, \Delta v
\\
conj\left(\tfrac{\partial u}{\partial y}\right)  \, \Delta u + conj\left(\tfrac{\partial v}{\partial y}\right)  \, \Delta v
\end{pmatrix}
=
\begin{pmatrix}
conj\left(\tfrac{\partial u}{\partial x}\right)  & conj\left(\tfrac{\partial v}{\partial x}\right)  \\
conj\left(\tfrac{\partial u}{\partial y}\right)  & conj\left(\tfrac{\partial v}{\partial y}\right)  \\
\end{pmatrix}
\begin{pmatrix}
\Delta u \\ \Delta v
\end{pmatrix}
```

> \end{pmatrix}

Reviewer 6: I think there's a missing minus sign here in the `∂/∂y` terms when we compare to https://github.com/JuliaDiff/ChainRulesCore.jl/pull/167/files#diff-a5f548edfc6b320f451aa0189937f0e8R138-R146. 

`imag(du_dz)` is equal to `-du_dy`, not `+du_dy`.

Reviewer 6: Actually, after checking my math, I think it's this section that's okay and `jacobian_via_rrule` that needs minus signs on the `imag` parts.

Reviewer 6: Actually, after checking my math, I think it's this section that's okay and `jacobian_via_rrule` that needs minus signs on the `imag` parts.

Reviewer 6: Okay, actually none of this is wrong, it's just that I think when @ettersi wrote 
```julia
_,du_dz = pullback( 1)
_,dv_dz = pullback(im)
```
he was referring to `du_dz` as `∂u/∂z = ∂u/∂x + i ∂u/∂y` rather than `∂u/∂z = ∂u/∂x - i ∂u/∂y`. 

Perhaps these should be renamed `du_dzconj` and `dv_dzconj` or something just in case someone takes that too literally? It's not a big deal though and maybe this is only confusing to me.

Reviewer: > Perhaps these should be renamed `du_dzconj` and `dv_dzconj` or something like
> 
> ```julia
> du_dx, du_dy = reim(pullback( 1)[2])
> dv_dx, du_dy = reim(pullback(im)[2])
> ```
> 
> just in case someone takes the `d_dz` too literally? It's not a big deal though and maybe this is only confusing to me.

I second the `reim` approach for clarity.

Author: Agreed, that was confusing. Thanks for spotting and fixing this!

.
```
The Jacobian of ``f:\mathbb{C} \to \mathbb{C}`` interpreted as a function ``\mathbb{R}^2 \to \mathbb{R}^2`` can hence be evaluated using either of the following functions.
```

> function jacobian_via_frule(f,z)
>     fz,df_dx = frule((Zero(), 1),f,z)
>     fz,df_dy = frule((Zero(),im),f,z)
>     return [
>         real(df_dx)  real(df_dy)
>         imag(df_dx)  imag(df_dy)
>     ]
> end

Reviewer 6: ```suggestion
function jacobian_via_frule(f,z)
    du_dx, dv_dx = reim(frule((Zero(), 1),f,z)[2])
    du_dy, dv_dy = reim(frule((Zero(),im),f,z)[2])
    return [
        du_dx  du_dy
        dv_dx  dv_dy
    ]
end
```

```
```

> function jacobian_via_rrule(f,z)
>     fz, pullback = rrule(f,z)
>     _,du_dz = pullback( 1)
>     _,dv_dz = pullback(im)
>     return [
>         real(du_dz)  imag(du_dz)
>         real(dv_dz)  imag(dv_dz)
>     ]
> end

Reviewer 6: ```suggestion
function jacobian_via_rrule(f,z)
    _, pullback = rrule(f,z)
    du_dx, du_dy = reim(pullback( 1)[2])
    dv_dx, dv_dy = reim(pullback(im)[2])
    return [
        du_dx  du_dy
        dv_dx  dv_dy
    ]
end
```
